### DISCUSSION TITLE
Serverpod auth example client can't reach server

### DISCUSSION THREAD

**Discussion by lukehutch** [serverpod/serverpod#541] [Q&A] [answered]
I copied code from `serverpod/examples/auth_example/auth_example_flutter/` to my own Flutter project. This project was able to successfully connect to the serverpod server in its module before I added this code.

However, the auth code does not seem to be able to contact the server. It seems to be using the wrong port (56888, not 8080):

```
I/flutter ( 4355): Failed call: serverpod_auth.google.authenticateWithServerAuthCode
I/flutter ( 4355): SocketException: Connection refused (OS Error: Connection refused, errno = 111), address = localhost, port = 56888
I/flutter ( 4355): serverpod_auth_google: SocketException: Connection refused (OS Error: Connection refused, errno = 111), address = localhost, port = 56888
I/flutter ( 4355): #0      _NativeSocket.startConnect (dart:io-patch/socket_patch.dart:682:35)
I/flutter ( 4355): #1      _RawSocket.startConnect (dart:io-patch/socket_patch.dart:1827:26)
I/flutter ( 4355): #2      RawSocket.startConnect (dart:io-patch/socket_patch.dart:27:23)
I/flutter ( 4355): #3      Socket._startConnect (dart:io-patch/socket_patch.dart:2048:22)
I/flutter ( 4355): #4      Socket.startConnect (dart:io/socket.dart:759:21)
I/flutter ( 4355): #5      _ConnectionTarget.connect (dart:_http/http_impl.dart:2453:20)
I/flutter ( 4355): #6      _HttpClient._getConnection.connect (dart:_http/http_impl.dart:2867:12)
I/flutter ( 4355): #7      _HttpClient._getConnection (dart:_http/http_impl.dart:2872:12)
I/flutter ( 4355): #8      _HttpClient._openUrl (dart:_http/http_impl.dart:2727:12)
I/flutter ( 4355): #9      _HttpClient.postUrl (dart:_http/http_impl.dart:2601:49)
I/flutter ( 4355): #10     ServerpodClient.callServerEndpoint
package:serverpod_client/src/serverpod_client_io.dart:71
I/flutter ( 4355): <asynchronous suspension>
I/flutter ( 4355): #11     signInWithGoogle
package:serverpod_auth_google_flutter/src/auth.dart:70
I/flutter ( 4355): <asynchronous suspension>
```

The client is set up as:

```
  client = Client(
    'http://localhost:8080/',
    authenticationKeyManager: FlutterAuthenticationKeyManager(),
  )..connectivityMonitor = FlutterConnectivityMonitor();
```

A similar error occurs for both Google auth and email auth.

Why is this trying to reach the server at port 56888?

**Update (1)**: I stepped through the code, and the actual URL being requested is `http://localhost:8080/serverpod_auth.google`. So The port number in the error message is wrong for some reason.

I think what is happening is that the endpoint `serverpod_auth/serverpod_auth_server/lib/src/endpoints/google_endpoint.dart` is not being registered with the server. How do I register this? I currently have the following in `pubspec.yaml`. What else is needed?

```
dependencies:
  serverpod: ^0.9.21
  serverpod_auth_server: ^0.9.21
```

**Update (2)**: The correct URL `http://localhost:8080/serverpod_auth.google` is being fetched by `serverpod_client_io.dart:71` as:

```
      var request = await _httpClient.postUrl(url);
```

However this throws an exception stating the wrong port number:

```
address: _InternetAddress (InternetAddress('127.0.0.1', IPv4))
message: "Connection refused"
osError: OSError (OS Error: Connection refused, errno = 111)
port: 55074
hashCode: 721684399
runtimeType: Type (SocketException)
```

The endpoint `serverpod_auth.google` is in fact being correctly registered in the server -- in `mypod_server/lib/src/generated/endpoints.dart` I see:

```
import 'package:serverpod_auth_server/module.dart' as _i3;
// ...
    modules['serverpod_auth'] = _i3.Endpoints()..initializeEndpoints(server);
```

If I visit `http://localhost:8080/serverpod_auth.google` in a web browser, I get a 400 error, so at least it's not giving `Connection refused`.

Any help debugging this would be greatly appreciated.

**Comment by vlidholt** [comment]
First, the port number that Flutter outputs is not the port number that it tries to connect to, but the inbound connection (which will be a random number). So that is not the issue.

The 400 error, represents access denied. You will get that when you connect to an endpoint you do not have access to. E.g. you will need to authenticate before accessing it.

Some things to check:
1. Did you run `serverpod generate`?
2. Have you added the database tables for the authentication module?
3. Are you connecting to your local server?

**Comment by lukehutch** [comment, marked as answer]
Sorry, answered in #544. Thanks.
